cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/profile/ProfileStore.java

```diff
@@ -82,94 +88,152 @@ public static void setupDatasets(DatasetFramework framework) throws IOException,
-    }, IOException.class);
+      getMDS(context).write(getRowKey(profileId), new ProfileMeta(profile, ProfileStatus.ENABLED));
+    });
   }
 
   /**
    * Add a profile if it does not exist in the store
    *
    * @param profileId the id of the profile to add
    * @param profile the information of the profile
-   * @throws IOException if there was an IO error adding the profile
    */
-  public void createIfNotExists(ProfileId profileId, Profile profile) throws IOException {
+  public void createIfNotExists(ProfileId profileId, Profile profile) {
     Transactionals.execute(transactional, context -> {
       MetadataStoreDataset mds = getMDS(context);
       MDSKey rowKey = getRowKey(profileId);
       Profile oldProfile = mds.get(rowKey, Profile.class);
       if (oldProfile != null) {
         return;
       }
-      mds.write(rowKey, profile);
-    }, IOException.class);
+      mds.write(rowKey, new ProfileMeta(profile, ProfileStatus.ENABLED));
+    });
   }
 
   /**
    * Deletes the profile from the profile store
    *
    * @param profileId the id of the profile to delete
-   * @throws IOException if there was an IO error deleting the profile
    * @throws NotFoundException if the profile is not found
    */
-  public void deleteProfile(ProfileId profileId) throws IOException, NotFoundException {
+  public void deleteProfile(ProfileId profileId) throws NotFoundException, BadRequestException {
     Transactionals.execute(transactional, context -> {
       MetadataStoreDataset mds = getMDS(context);
       MDSKey rowKey = getRowKey(profileId);
-      Profile value = getMDS(context).get(rowKey, Profile.class);
+      ProfileMeta value = mds.get(rowKey, ProfileMeta.class);
       if (value == null) {
         throw new NotFoundException(profileId);
       }
+      // TODO: https://issues.cask.co/browse/CDAP-13514, check for associations of the profile before deleting
+      if (value.getStatus() == ProfileStatus.ENABLED) {
+        throw new BadRequestException(String.format("Profile %s in namespace %s is currently enabled. A profile can " +
+                                                      "only be deleted if it is disabled",
+                                                    profileId.getProfile(), profileId.getNamespace()));
+      }
       mds.delete(getRowKey(profileId));
-    }, IOException.class, NotFoundException.class);
+    }, NotFoundException.class, BadRequestException.class);
   }
 
   /**
    * Delete all profiles in a given namespace.
    *
    * @param namespaceId the id of the namespace
-   * @throws IOException if there was an IO error deleting the profiles
    */
   public void deleteAllProfiles(NamespaceId namespaceId) throws IOException {
     Transactionals.execute(transactional, context -> {
       getMDS(context).deleteAll(getRowKey(namespaceId));
-    }, IOException.class);
+    });
+  }
+
+  public ProfileStatus getProfileStatus(ProfileId profileId) throws NotFoundException {
```

Comment:
don't think there is value in having a separate method for this.

Reply:
removed